### PR TITLE
Some simplifier rules

### PR DESCRIPTION
**Pull request by msooseth**
## Description

Lots of simplifier rules -- a bit too much, in fact, because the entire contract now disappears sometimes :) The SLT I'm not sure about. All of it needs a review. SLT clearly needs more love, even if correct. Negative to positive comparison is not working, and also, it's probably messy? Maybe wrong? Can you check?

### CODE REVIEW

File: src/hevm/src/EVM/SymExec.hs [hunk 1]
```diff
@@ -335,9 +335,15 @@ simplify e = if (mapExpr go e == e)
...
 
     -- simplify buffers
+    go o@(ReadWord a (WriteWord b c _))
+      | a == b = c
+      | otherwise = o
     go o@(ReadWord (Lit _) _) = Expr.simplifyReads o
     go o@(ReadByte (Lit _) _) = Expr.simplifyReads o
 
+    go o@(BufLength (Fact _ (WriteByte _ _ _))) = Lit 0x1
+    go o@(BufLength (Fact _ (WriteWord _ _ _))) = Lit 0x4
```
Comment (d-xo) [inline src/hevm/src/EVM/SymExec.hs:345]: why?

File: src/hevm/src/EVM/SymExec.hs [hunk 2]
```diff
@@ -335,9 +335,15 @@ simplify e = if (mapExpr go e == e)
     go (CopySlice (Lit 0x0) (Lit 0x0) (Lit 0x0) _ dst) = dst
 
     -- simplify buffers
+    go o@(ReadWord a (WriteWord b c _))
```
Comment (d-xo) [inline src/hevm/src/EVM/SymExec.hs:338]: Not necessarily opposed to this rule, but curious it when we have Expr.readWord?

File: src/hevm/src/EVM/SymExec.hs [hunk 3]
```diff
@@ -353,12 +359,32 @@ simplify e = if (mapExpr go e == e)
...
+      | a == (Lit 0) = b
+      | otherwise = o
+    go o@(EVM.Types.LT (Lit a) (Lit b))
+      | (a < b) = Lit 1
+      | otherwise = Lit 0
+    go o@(SLT (Lit a) (Lit b))
+      | (a < 128)  && (b < 128)  && (a < b)  = Lit 1
+      | (a < 128)  && (b < 128)  && (a >= b) = Lit 0
+      | (a >= 128) && (b >= 128) && (a < b)  = Lit 0
+      | (a >= 128) && (b >= 128) && (a >= b) = Lit 1
```
Comment (d-xo) [inline src/hevm/src/EVM/SymExec.hs:382]: why?

File: src/hevm/src/EVM/SymExec.hs (hunk 2 above)
Comment (msooseth) [reply src/hevm/src/EVM/SymExec.hs:338]: The function selection(?) seems to produce this:

```
            (ITE (Eq (ReadWord (Lit 0x4) (Fact (PGEq (BufLength (WriteWord (Lit 0x4) (Fact (PAnd (PGEq (Var "arg1") (Lit 0x0)) (PLEq (Var "arg1") (Lit 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff))) (Var "arg1")) (WriteByte (Lit 0x0) (LitByte 182) (WriteByte (Lit 0x1) (LitByte 181) (WriteByte (Lit 0x2) (LitByte 95) (WriteByte (Lit 0x3) (LitByte 37) (AbstractBuf "txdata"))))))) (Lit 0x24)) (WriteWord (Lit 0x4) (Fact (PAnd (PGEq (Var "arg1") (Lit 0x0)) (PLEq (Var "arg1") (Lit 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff))) (Var "arg1")) (WriteByte (Lit 0x0) (LitByte 182) (WriteByte (Lit 0x1) (LitByte 181) (WriteByte (Lit 0x2) (LitByte 95) (WriteByte (Lit 0x3) (LitByte 37) (AbstractBuf "txdata")))))))) (Lit 0x0))
```

**Pull request closed without merging**